File: algorithms-DS/quick-sort/java_quick_sort/QuickSortFromBothEnds.java
import java.util.List;
import java.util.ArrayList;

public class QuickSortFromBothEnds {

	private List<Double> array;

	QuickSortFromBothEnds(List<Double> arr) {
		this.array = arr;
	}

	public void sort() {
		this.sort(this.array, 0, this.array.size() - 1);
	}

	public void sort(List<Double> arr) {
		this.sort(arr, 0, arr.size() - 1);
	}

	public void sort(List<Double> arr,  Integer floor, Integer ceil) {
		if(floor <= ceil) {
			Integer pivot = partition(arr, floor, ceil);

			sort(arr, floor, pivot - 1);
			sort(arr, pivot + 1, ceil);
		}
	}

	private Integer partition(List<Double> arr, Integer floor, Integer ceil){
		Integer leftMarker	= floor;
		Integer rightMarker 	= ceil - 1;

		while(leftMarker <= rightMarker){

			this.swap(arr, leftMarker, rightMarker);

			while((leftMarker <= rightMarker) &&
			      (arr.get(leftMarker) <= arr.get(ceil)))

				leftMarker += 1;

			while((leftMarker <= rightMarker) &&
			      (arr.get(rightMarker) >= arr.get(ceil)))

				rightMarker -= 1;

		}
		swap(arr, leftMarker, ceil);
		return leftMarker;
	}

	private void swap(List<Double> arr, Integer pos1, Integer pos2){
		Double tmp = arr.get(pos1);
		arr.set(pos1, arr.get(pos2));
		arr.set(pos2, tmp);
	}

	public void printArr(){
		System.out.println(this.array);
	}
}

File: algorithms-DS/quick-sort/python_quick_sort/quick_sort_from_both_ends.py
def quick_sort(arr, floor, ceil):
    if(floor < ceil):
        pivot = partition(arr, floor, ceil)

        quick_sort(arr, floor, pivot - 1)
        quick_sort(arr, pivot + 1, ceil)

def partition(arr, floor, ceil):
    left_marker = floor
    right_marker = ceil - 1

    print("in : ", arr, floor, ceil)
    while(left_marker <= right_marker):
        
        if((arr[left_marker] >= arr[ceil]) and
           (arr[right_marker] < arr[ceil])):

            swap(arr, left_marker, right_marker)

        if(arr[left_marker] < arr[ceil]):
            left_marker += 1
        if(arr[right_marker] >= arr[ceil]):
            right_marker -= 1

    swap(arr, left_marker, ceil)
    return left_marker

def swap(arr, pos1, pos2):
    tmp = arr[pos1]
    arr[pos1] = arr[pos2]
    arr[pos2] = tmp
